### PR TITLE
:sparkles: 實作一個人同時只能加入一個房間。

### LINKED ISSUE #61: 一人只能進入一間房
# 一人只能進入一間房 (待討論)
```
Given：
「房間 01」& 「房間 02」 在房間列表裡
「房間 01」裡有 **使用者 A**
「房間 02」裡有 **使用者 B**
When：**使用者 B** 加入「房間 01」
Then：Error，一人只能進入一間房
```

---
# 待討論的點:
這裡有第二種作法，不回傳 Error，直接離開原有的房間，並加入新房間
```
Given：
「房間 01」 &「房間 02」在房間列表裡
「房間 01」裡有 **使用者 A**
「房間 02」裡有 **使用者 B**
When：**使用者 B** 加入「房間 01」
Then：
**使用者 B** 離開「房間 02」
**使用者 B** 加入「房間 01」
「房間 02」因為沒有人了，從房間列表移除
「房間 01」依然在房間列表裡
「房間 01」裡有 **使用者 A & B**
```

### PR DESCRIPTION
**Pull request by frankvicky**
## Why need this change? / Root cause: 
- 一個人同時只能加入一個房間。
## Changes made:
- JoinRoomUsecase
## Test Scope / Change impact:
- RoomControllerTest
## Issue
- resolved #61

### CODE REVIEW

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 1]
```diff
@@ -18,11 +18,20 @@ class JoinRoomUsecase(
...
+        val room = findRoomById(Room.Id(roomId))
+        validateUserJoinSingleRoom(userId)
+        room.run {
+            validateRoomPassword(password)
+            validateFullRoom()
+            joinPlayer(userId)
+        }
+    }
+
+    private fun validateUserJoinSingleRoom(userId: String) {
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:31]: `validateUserJoinedRoom`

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 2]
```diff
@@ -18,11 +18,20 @@ class JoinRoomUsecase(
...
+        val isJoined = roomRepository.hasPlayerJoinedRoom(User.Id(userId))
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:32]: `hasJoined`

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 3]
```diff
@@ -18,11 +18,20 @@ class JoinRoomUsecase(
...
+        if (isJoined) {
+            throw PlatformException("Can only join a single room at same time")
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:34]: `Player($userId) has joined another room.`

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/dao/RoomDAO.kt [hunk 4]
```diff
@@ -6,10 +6,12 @@ import org.springframework.data.mongodb.repository.MongoRepository
 import org.springframework.stereotype.Repository
 import tw.waterballsa.gaas.domain.Room
 import tw.waterballsa.gaas.spring.repositories.data.RoomData
+import tw.waterballsa.gaas.spring.repositories.data.RoomData.PlayerData
 
 @Repository
 interface RoomDAO : MongoRepository<RoomData, String> {
     fun existsByHostId(hostId: String): Boolean
     fun findByStatus(status: Room.Status, pageable: Pageable): Page<RoomData>
+    fun existsByPlayersContaining(players: List<PlayerData>): Boolean
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/dao/RoomDAO.kt:15]: `Collection`

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/data/RoomData.kt [hunk 5]
```diff
@@ -36,10 +36,10 @@ class RoomData(
             password = password
         )
 
-    class PlayerData(
+    data class PlayerData(
         val id: String,
         val nickname: String,
-        private val readiness: Boolean
+        val readiness: Boolean
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/data/RoomData.kt:42]: Ｑ：這不是 private 的原因是？

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt [hunk 6]
```diff
@@ -284,6 +286,28 @@ class RoomControllerTest @Autowired constructor(
...
 
+    @Test
+    @DisplayName(
+        """
+        Given: Player A has already joined a room B
+        When: Player A join another room C
+        Then: Should fail
+    """
+    )
+    fun testUserJoinTwoRoomsAtSameTime() {
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt:297]: `testUserJoinAnotherRoom()`

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt [hunk 7]
```diff
@@ -284,6 +286,28 @@ class RoomControllerTest @Autowired constructor(
             .thenPlayerShouldBeNotInRoomAndHostIsChanged(host)
     }
 
+    @Test
+    @DisplayName(
+        """
+        Given: Player A has already joined a room B
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/RoomControllerTest.kt:292]: `the room B`

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/data/RoomData.kt (hunk 5 above)
Comment (frankvicky) [reply spring/src/main/kotlin/tw/waterballsa/gaas/spring/repositories/data/RoomData.kt:42]: 誤刪

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Pull request merged**